components/automatic-grid/script.js

```diff
@@ -0,0 +1,244 @@
+    }
+    
+    /**
+     * Gets the first empty cell of the cells array
+     * @returns {HTMLElement}
+     */
+    get firstEmptyCell() {
+        return this._cells.find((cell) => cell.childElementCount === 0);
+    }
+
+    constructor() {
+        super();
+        // use the template to create the base grid
+        this.template = template;
+
+        this.dragStart = this.dragStart.bind(this);
+        this.dragMove = this.dragMove.bind(this);
+        this.dragEnd = this.dragEnd.bind(this);
+    }
+
+    /**
+     * Called when the element was attached to the DOM.
+     */
+    connectedCallback() {
+        //Get the number of columns. Max number of columns is 12
+        this._columns = this.setColumns();
+        //Get the number of rows. No max number of rows. If no row number is passed it defaults to 5
+        this._rows = this.getAttribute('rows') || 5;
+
+        //The number of cells in the grid
+        this._numberOfCells = this._rows * this._columns;
+
+        //If the items in the cells can be drag and dropped
+        this._draggable = this.hasAttribute('draggable');
+
+        //The array of cells
+        this._cells = [];
+
+        // Load the template
+        components
+            .loadResource(this)
+            .then((result) => {
+                this.template = result.template;
+                //render the template
+                components.renderOnce(this);
+
+                //Get the grid container, where all of the cells will be
+                this._gridContainer = this.querySelector('.guic-automatic-grid-container');
+
+                this.buildGrid();
+                this.addItemsToCells();
+            })
+            .catch((err) => console.error(err));
+    }
+
+    /**
+     * Builds the grid, by adding rows and cells.
+     */
+    buildGrid() {
+        //Loop through all of the cells
+        for (let i = 0; i < this._numberOfCells; i++) {
+            //Since we are using the grid component, we need to make a row element for each grid row. When the index of the cell is the begining of each column we create a new row.
+            if (i % this._columns === 0) {
+                const row = document.createElement('div');
+                row.classList.add('guic-row');
+
+                this._gridContainer.appendChild(row);
+            }
+            const element = document.createElement('div');
+
+            element.classList.add('automatic-grid-cell');
+            element.classList.add(`guic-col-${AutomaticGrid.formatFloat(maxColumns / this._columns)}`); //We add the grid class to each cell based on the number of columns
+
+            element.dataset.col = (i % this._columns) + 1; //We set the cell column so that it's available if we need it later. +1 so that columns don't start from 0
+            element.dataset.row = Math.floor(i / this._columns) + 1; //We set the cell row so that it's available if we need it later. +1 so that rows don't start from 0
```

Comment:
We can avoid doing these calculations if we use nested for loop - rows by columns. The overall complexity of the algorithm might be higher, but the remainder operator compensates enough.
P.S. - just an observation, don't feel obligated to change it. It's fine and it works.